lib/chibi/syntax-case.scm

```diff
@@ -0,0 +1,323 @@
+;; Written by Marc Nieper-Wißkirchen
+
+;; TODO: make-variable-transformer and identifier-syntax.
+
+;; TODO: make-synthetic-identifier should return a truly unique (that
+;; is not free-identifier=? to any other) identifier.
+
+;; TODO: Consecutive ellipses in syntax templates.
+
+;; TODO: Write many more tests.
+
+(define-syntax define-pattern-variable
+  (er-macro-transformer
+   (lambda (expr rename compare)
+     (let ((id (cadr expr))
+	   (binding (cddr expr)))
+       (let ((mac (cdr (env-cell (current-usage-environment) id))))
+	 (macro-aux-set! mac binding))
+       `(,(rename 'begin))))))
+
+(define (make-pattern-variable pvar)
+  (lambda (expr)
+    (error "reference to pattern variable outside syntax" pvar)))
+
+(define (pattern-variable x)
+  (let ((cell (env-cell (current-usage-environment) x)))
+    (and cell (macro? (cdr cell)) (macro-aux (cdr cell)))))
+
+(define (rename id)
+  ((current-renamer) id))
+
+(define (ellipsis? id)
+  (free-identifier=? id (rename '...)))
+
+(define bound-identifier=?
+  (lambda (x y)
+    (eq? x y)))
+
+(define (syntax-transformer level)
+  (er-macro-transformer
+   (lambda (expr rename compare)
+     (let*-values (((out envs)
+		    (gen-template (cadr expr) '() ellipsis? level)))
+       out))))
+
+(define (syntax->datum stx)
+  (strip-syntactic-closures stx))
+
+(define (symbol->identifier id symbol)
+  (if (symbol? id)
+      symbol
+      ((syntactic-closure-rename id)
+       symbol)))
+
+;; TODO: Handle cycles in datum.
+(define (datum->syntax id datum)
```

Comment:
I think it's worth having datum->syntax in the core.

Reply:
Sounds good. I'm moving it back to init-7.scm and re-export it from `(chibi syntax-case)` (so that all relevant bindings can also be found in one place for programs that do not want to import the whole of `(chibi)`).

In the long run, one may think to move even `syntax` and `syntax-case` into the core. Not because they need to be exported by `(chibi)` but because the code needed to implement `syntax-rules` in the core is roughly the same code needed to implement `syntax-case` and `syntax` and we could avoid code-doubling.

By defining `syntax`/`quasisyntax` first, I was able to use it in subsequent macros (like `syntax-case`) making their code much easier to read and write (because I didn't have to use `rename` all over the place). Some macros in `init-7.scm` may also benefit from this.